### PR TITLE
Benchmarking dependency adoption

### LINKED ISSUE #28: Benchmarking design decisions - an example with `cli::cli_warn()`
I discovered that `cli::cli_warn()` is a slower alternative to `base::warning()` when throwing warnings. 

I have drafted a rough blog post on this by relating it to the use of benchmarking to weigh design decisions. 

Do let me know what you think.

### PR DESCRIPTION
**Pull request by jamesmbaazam**
In deciding between `cli::cli_warn()` and `base::warning()` in a refactoring exercise, I discovered that the former was an order of magnitude slower than the latter. This was interesting as I first had an intuition and explored it through a benchmarking exercise. This has inspired me to draft a blog post by relating it to the use of benchmarking to make data-driven design decisions in R package development.

This PR closes #28.

### CODE REVIEW

File: posts/benchmarking_design_decisions/index.qmd [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+These design decisions often result in taking on dependencies. Dependencies are
+external packages that enable a function to work properly. For example, the
+`tidyverse` is a collection of packages that are often used together to
+make data analysis easier. The `tidyverse` is a dependency for many
+packages that are built on top of it.
+
+Dependencies should not be adopted blindly without carefully considering the
+overhead they introduce. Overhead here refers to the additional work that a
+package must do to make the dependency work. In this article, we will
+demonstrate how to use (micro)benchmarking[^1] to guide the decision to take on
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:36]: Looks like you opted to use the bench package in the end so this note should probably be removed.

File: posts/benchmarking_design_decisions/index.qmd [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+dependencies.
+
+For this demonstration, we will look at condition handling and signalling,
+that is, what to do when problems arise during code execution [^2].
+Condition signalling in R is classified into errors and warnings. Errors refer
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:41]: We maybe don't need to go into a technical discussion here but just flagging that this is not entirely accurate. Messages are conditions as well and R allows you to create conditions of any class.

File: posts/benchmarking_design_decisions/index.qmd [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+need to halt the process.
+
+```{r}
+c(1:2) > c(1:3)
+```
+
+Base R ships with functionalities for condition handling and signalling. There is
+`stop()` for signaling errors and `warning()` for throwing warnings. These base R
+functions are, however, are not flexible enough for advanced formatting: colour,
+progress bar, contextual pluralisation, and so forth.
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:59]: Nitpicking here but: base R is flexible enough for this and this is precisely how cli manages to build on top of it to add extra features.

What about something like:

```suggestion
`stop()` for signaling errors and `warning()` for throwing warnings. These base R
functions do not include desirable advanced formatting features: colour,
progress bar, contextual pluralisation, and so forth.
```

File: posts/benchmarking_design_decisions/index.qmd [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+Package developers often have to adopt external dependencies for flexibility
+in condition signalling. The [cli](https://cli.r-lib.org/articles/semantic-cli.html) package is well designed for this. `cli` offers many advanced features
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:62]:
```suggestion
Package developers often have to adopt external dependencies for feature-rich condition signalling. The [cli](https://cli.r-lib.org/articles/semantic-cli.html) package is well designed for this. `cli` offers many advanced features
```

File: posts/benchmarking_design_decisions/index.qmd [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+that will not be discussed here, but of note is its `cli_abort()` and
+`cli_warn()` functions for signalling errors and warnings respectfully.
+
+`cli` is a great candidate for flexibility in condition signalling but one
+might naturally wonder if this comes at a cost. Here,
+we will investigate this through a benchmarking exercise against base R's
+`warning()` function. The results of this exercise will provide us with insights
+on whether its worth taking on `cli` for its flexibility as a trade-off for its
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:70]:
```suggestion
on whether its worth taking on `cli` for its features as a trade-off for its
```

File: posts/benchmarking_design_decisions/index.qmd [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+in a data.frame.
+
+```{r warning=FALSE}
+bnmark_list <- warnings |>
+  map(
+    \(x) mark(
+      "cli::cli_warn()" = warn_Ntimes(x, pkg = "cli"),
+      "base::warning()" = warn_Ntimes(x, pkg = "base")
+    )
+  )
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:136]: Since you're using bench, it's probably simpler to use `press()` directly. It should be something like:

```r
bench::press(
  fct = c("cli", "base"),
  times = c(1, 5, 10, 15, 20, 100),
  warn_Ntimes(times, fct)
)
```

https://bench.r-lib.org/reference/press.html

File: posts/benchmarking_design_decisions/index.qmd [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+## Conclusion
+
+In designing R package infrastructure with "heavy" dependencies, it might
+sometimes be necessary to test your intuition about their overhead with
+experiments. Here, we have demonstrated how benchmarking is one way to achieve
+this. We show how a simple decision to use `cli::cli_warn()` to handle
+warnings could come at the cost of a tiny loss in speed, which is worth
+considering in its context.
+
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:205]: Would it make sense to link to https://r-pkgs.org/dependencies-mindset-background.html#sec-dependencies-pros-cons for a longer discussion on why every dependency taking decision is unique and how to take context into account?

File: posts/benchmarking_design_decisions/index.qmd (hunk 1 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:36]: Not sure I understand. I'm not yet referring to any package in this paragraph.

File: posts/benchmarking_design_decisions/index.qmd (hunk 2 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:41]: Thanks. I've added that.

File: posts/benchmarking_design_decisions/index.qmd (hunk 3 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:59]: Awesome. Thanks.

File: posts/benchmarking_design_decisions/index.qmd (hunk 6 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:135]: Wow, that's very convenient. Thanks.

File: posts/benchmarking_design_decisions/index.qmd (hunk 7 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:222]: I've linked it now.

File: posts/benchmarking_design_decisions/index.qmd [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+)
+```
+
+R package developers are constantly faced with design decisions about user
+friendliness. Such design decisions often include checking that inputs are
+rightly specified and returning feedback early, deciding on how to deal with
+issues arising during code execution within a function, and how to
+structure outputs.
+
+These design decisions often result in taking on dependencies. Dependencies are
```
Comment (joshwlambert) [inline posts/benchmarking_design_decisions/index.qmd:28]: Not sure I agree with this definition of what a dependency is. A function could work properly without calling code from an external package, but it might be better to not rewrite that code, or the dependency could provide a more feature-rich function than base R. I would go with the more general definition used here: https://r-pkgs.org/dependencies-mindset-background.html

File: posts/benchmarking_design_decisions/index.qmd [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+external packages that enable a function to work properly. For example, the
+`tidyverse` is a collection of packages that are often used together to
```
Comment (joshwlambert) [inline posts/benchmarking_design_decisions/index.qmd:30]: I like the use of the tidyverse as an example as many readers will be familiar with it. However, it might not be the best example for this as it is not a dependency that in itself provides any functionality, but rather bundles functionality from other packages. I think either way the reader will understand what is being explained. 

Might also be good to hyperlink one of the "`tidyverse`" to their website.

File: posts/benchmarking_design_decisions/index.qmd [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+make data analysis easier. The `tidyverse` is a dependency for many
+packages that are built on top of it.
+
+Dependencies should only be adopted after carefully considering their pros and
+cons. In this article, we will demonstrate how to use (micro)benchmarking[^1]
+to guide the decision to take on dependencies that might lead to slower
+processes. In data analytics pipelines, speed is often a key consideration,
```
Comment (joshwlambert) [inline posts/benchmarking_design_decisions/index.qmd:37]:
```suggestion
cons. In this article, we will demonstrate how to use (micro)benchmarking[^1]
to guide decision-making on dependencies that might lead to slower
processes. In data analytics pipelines, speed is often a key consideration,
```

File: posts/benchmarking_design_decisions/index.qmd (hunk 1 above)
Comment (Bisaloo) [reply posts/benchmarking_design_decisions/index.qmd]: Oh, I misunderstood you intent then. I took this was a foreshadowing of the fact you were going to use the microbenchmark package.
I now understand was to refer to the concept. Feel free to disregard this comment then.

File: posts/benchmarking_design_decisions/index.qmd [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+need to halt the process. For example, when you compare two vectors of
+different lengths, R will throw a warning but will still return a result.
+
+```{r}
+c(1:2) > c(1:3)
+```
+
+Messages are used to provide useful information about processes. For
+example, when you load a package, R will throw a message to indicate that the
+package has been loaded.
```
Comment (joshwlambert) [inline posts/benchmarking_design_decisions/index.qmd:63]: I think this only happens when there are namespace conflicts, and otherwise the package loads silently (assuming the package author didn't write a message to be printed in `.onLoad()`).

File: posts/benchmarking_design_decisions/index.qmd [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+functions are, however, not flexible enough for advanced
+formatting: colour, progress bar, contextual pluralisation, and so forth.
+
+Package developers often have to adopt external dependencies for flexibility
+in condition signalling. The [cli](https://cli.r-lib.org/articles/semantic-cli.html) package is well designed for this. `cli` offers many advanced features
+that will not be discussed here, but of note is its `cli_abort()` and
+`cli_warn()` functions for signalling errors and warnings respectfully.
+
+`cli` is a great candidate for flexibility in condition signalling but one
+might naturally wonder if this comes at a cost. Here,
```
Comment (joshwlambert) [inline posts/benchmarking_design_decisions/index.qmd:81]: Before reading this post, I would have naively assumed that {cli} was performant enough not to care about understanding/quantifying its speed. Therefore, it might be worth mentioning that benchmarking can be used to check whether a introduced function from an external package slows down the code even when it is not expected. This point about unexpected slowdowns/bottlenecks might be better placed in the conclusion.

File: posts/benchmarking_design_decisions/index.qmd [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+bnmark_relative_speeds
+```
+
+As we can see in the new column `relative_speed`, `cli` is consistently
+slower than base R and this seems not to
+scale linearly with the number of warnings thrown. Other benchmarking
+experiments
+with `cli` have revealed linear relationships between its speed and the number
+of times it invokes other functions [^3]. The developers of `cli` have also
+conducted benchmarks of the `ansi_` family of functions in `cli` in comparison
```
Comment (joshwlambert) [inline posts/benchmarking_design_decisions/index.qmd:194]: Could add an asterisk (wildcard character) to `ansi_` (`ansi_*`) to show that this is a prefix for a family of functions.

File: posts/benchmarking_design_decisions/index.qmd [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+to base R and the `fansi` package. They find that `cli` is consistently slower
+than base R, which corroborates the results of our experiment here.
+Their benchmarks are available in the cli documentation [^4]. So, should we
+be worried about the speed of `cli`? Well, it depends on the context. The
+"R Packages" book by Hadley Wickham and Jenny Bryan has a section on
+this where they suggest approaching such a decision from a holistic, balanced,
+and quantitative approach[^5]. We'll leave the reader to make their own decision
```
Comment (joshwlambert) [inline posts/benchmarking_design_decisions/index.qmd:201]: It might be worth fully citing the R packages book here. I don't know if we have a policy/preference on citations vs footnotes. I think citing would be a more clear sign of attribution for a published piece of work.

@Bisaloo any preferences?

**Comment by joshwlambert** [comment]
@jamesmbaazam nice work on this post! I enjoyed reading it and it gives a clear overview of the utility of benchmarking when making design decisions. 

It is a topic that has wide-reaching applications, but I like how it is self-contained and focuses on signal/error handling. There is plenty of possibility for a follow-up on benchmarking for design decisions around data structures and performing iterative tasks.

File: posts/benchmarking_design_decisions/index.qmd [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+* [rbenchmark](https://code.google.com/archive/p/rbenchmark/): an R package
+  for benchmarking R code.
+* [tictok](https://github.com/collectivemedia/tictoc): an R package to time
+R functions
+* [touchstone](https://github.com/lorenzwalthert/touchstone/tree/main): an R
+  package for
+  benchmarking of pull requests with statistical confidence.
+
+[^1]: Benchmark (Wikipedia): <https://en.wikipedia.org/wiki/Benchmark_(computing)>
+[^2]:  Hadley Wickham's Advanced R chapter on Exceptions and Debugging: <http://adv-r.had.co.nz/Exceptions-Debugging.html>
```
Comment (joshwlambert) [inline posts/benchmarking_design_decisions/index.qmd:241]: I think this links to the old version of Advanced R. I'm not sure how much the content differs between versions, but it might be worth linking to the newer version of book: https://adv-r.hadley.nz/conditions.html. (Please check the new version contain the relevant info).

File: posts/benchmarking_design_decisions/index.qmd [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+
+## Results
+
+It's time to explore the results of the data generated. We'll start by making
+a boxplot of the median run times for the different scenarios.
+
+Drum roll please...
+
+```{r}
+autoplot(
```
[the remark this answers is not in the document]
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:161]: The default is 10k times. I thought to mention it but also wasn't sure if it was relevant. I was surprised at the dispersion in the base R results as well. Not sure what would explain that.

File: posts/benchmarking_design_decisions/index.qmd (hunk 11 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:63]: True. I'll make the example more specific to this package.

File: posts/benchmarking_design_decisions/index.qmd [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+    ) |> 
+  mutate(
+    relative_speed = cli/base
+    )
+bnmark_relative_speeds
+```
+
+As we can see in the new column `relative_speed`, `cli` is consistently
+slower than base R and this seems not to
+scale linearly with the number of warnings thrown. Other benchmarking
```
Comment (pratikunterwegs) [inline posts/benchmarking_design_decisions/index.qmd:190]: The intention behind this statement wasn't clear to me - my expectation was that both {cli} and {base} times would scale linearly with $N$ warnings, which is met. If both scale roughly linearly, the relative difference should be roughly constant too as it is here. Is there a reason to expect otherwise?

It would be an issue if the relative difference _did_ scale, making {cli} unsuitable for high-performance applications with high-frequency output (e.g. perhaps some text-based interface for live monitoring?).

I would rephrase this to make it clear that there's an overhead to using {cli}, but that it is predictable, so there are unlikely to be nasty surprises (might be worth running for 10K warnings to be really sure).

**Review by pratikunterwegs — COMMENTED** [review]
Thanks for finishing this up @jamesmbaazam. I think it's a valuable topic to address, especially as some Epiverse packages have chosen to use {cli} without considering the performance costs (mea culpa).

That said, I think the post tries to combine two topics - an earlier one on taking on dependencies that improve user friendliness (cf. the opening sentence), and a more recent one on benchmarking dependencies generally - while sticking with the same example. {cli} is a good example of the former, but not the latter, IMO. The latter case would be better served with a more consequential example - the classic one would be {data.table} vs {dplyr} etc., but this has been covered at length elsewhere already.

I would suggest focusing on the trade-off between aesthetics/ergonomics/design and performance, using {cli} as an example. You would need to look at some literature on the design of interfaces and how it affects user experience and user performance, and how using {cli} features can help with that. One example might be using formatting to convey semantic meaning in messages or in print methods. You could then lay out the performance costs, and suggest ways to minimise them.

Here's an example of a hybrid warning where {cli} is used to generate coloured text for semantic meaning, but the ANSI string is passed to a base warning.

``` r
bench::mark(cli = cli::cli_warn("some text"), base = warning("some text"), 
    hybrid = warning(cli::col_red("some text")), check = F)
... <LOT OF WARNINGS>
#> # A tibble: 3 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 cli          2.54ms   2.92ms      292.     1.9MB     24.7
#> 2 base        28.62µs  61.42µs    15700.    4.54KB     21.5
#> 3 hybrid      98.65µs 123.74µs     8021.   22.08KB     23.9
```

<sup>Created on 2023-10-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Hope this helps, happy to take another look if you end up editing the post.

**Comment by jamesmbaazam** [comment]
> That said, I think the post tries to combine two topics - an earlier one on taking on dependencies that improve user friendliness (cf. the opening sentence), and a more recent one on benchmarking dependencies generally - while sticking with the same example. {cli} is a good example of the former, but not the latter, IMO. The latter case would be better served with a more consequential example - the classic one would be {data.table} vs {dplyr} etc., but this has been covered at length elsewhere already.

I think there's a logical link between the two issues you've outlined. Here, I argue that if you want to take on a dependency to improve user-friendliness, be sure to benchmark it to see if the tradeoffs between the features it introduces and speed loss are worth it. I agree `cli` is not a good example of the latter, if considered in its absolute sense but not in the context of this article. In general, everyone thinks about benchmarking only when it'll bring considerable gains in speed. By way of this article, I'm arguing that context matters. In the context of printing, if the process takes more than 2 blinks of the eye, that's too long. 

> I would suggest focusing on the trade-off between aesthetics/ergonomics/design and performance, using {cli} as an 
> example. You would need to look at some literature on the design of interfaces and how it affects user experience and user > performance, and how using {cli} features can help with that. 

Your proposed topic is a good candidate for another blog post. I did do a lot of research in trying to align this post with UX and UI design but that would have taken a detour from the core of the message, which is the utility of benchmarking.

**Comment by pratikunterwegs** [comment]
> I think there's a logical link between the two issues you've outlined. Here, I argue that if you want to take on a dependency to improve user-friendliness, be sure to benchmark it to see if the tradeoffs between the features it introduces and speed loss are worth it. I agree `cli` is not a good example of the latter, if considered in its absolute sense but not in the context of this article. In general, everyone thinks about benchmarking only when it'll bring considerable gains in speed. By way of this article, I'm arguing that context matters. In the context of printing, if the process takes more than 2 blinks of the eye, that's too long.

They're linked because one is a subset of the other, and I'm suggesting to tighten the scope to formatting/user-friendliness dependencies. By giving the example of {tidyverse}, the scope expands substantially to deps which are more important to performance than {cli}. Looking at the benchmarks, {cli} takes < 0.5s for 100 warnings, which isn't terrible.

Not to continue {cli} critcism, but perhaps it might be good to add in the post _why_ the performance loss occurs. Is it an issue with representing the string, or something else? I would also like to see the benefits discussed a bit more, basically any of the inline markup functionality.

**Comment by pratikunterwegs** [comment]
One point that I noticed but set aside initially was that a large number of files are changed in this PR, including CSS files. Would this be fixed by rebasing on `main`, or has a directory that should have been ignored leaked in? See https://github.com/epiverse-trace/epiverse-trace.github.io/pull/125/ for reference, where only a single file is changed (added).

File: posts/benchmarking_design_decisions/index.qmd (hunk 8 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:28]: Thanks. I've revised it here d08b3d6254cbb31a44d4e38ad2fd42c617d0bd80.

File: posts/benchmarking_design_decisions/index.qmd (hunk 9 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:30]: Yeah, I initially wanted to leave it as the generic bundle but I've narrowed it down to one example using ggplot2 here d08b3d6254cbb31a44d4e38ad2fd42c617d0bd80.

File: posts/benchmarking_design_decisions/index.qmd (hunk 11 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:63]: Thanks @pratikunterwegs. I'll add the qmd style option in the chunk.

File: posts/benchmarking_design_decisions/index.qmd (hunk 12 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:81]: Yes, that makes sense. I've made the change here a2a5ddf1625a16ddbb03aacdf1ffae0dc55635a6.

File: posts/benchmarking_design_decisions/index.qmd [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+  warning_msg <- "x must be an integer"
+  switch(pkg,
+    base = for (i in 1:n) {
+      warning(warning_msg)
+    },
+    cli = for (i in 1:n) {
+      cli_warn(warning_msg)
+    }
+  )
+}
```
[the remark this answers is not in the document]
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:122]: Always good to catch out these patterns. Thanks, Change made here 468f20252d24d68fcdce8c2adc3c83bbd6ede74a.

File: posts/benchmarking_design_decisions/index.qmd (hunk 13 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:194]: Thanks. I've made that change.

File: posts/benchmarking_design_decisions/index.qmd (hunk 14 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:201]: That's true.

Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:201]: I've switched to using citations for books.

File: posts/benchmarking_design_decisions/index.qmd (hunk 17 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd:190]: I've removed that section as it wasn't entirely useful. See 88c704d61df1a6d0fba81ab26c0aed018418795b.

**Comment by jamesmbaazam** [comment]
Hi @Bisaloo, how can I fix this CI issue?

**Comment by Bisaloo** [comment]
Hi James, I have deleted the extra files from this PR in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/36/commits/acf2ede453b5ea607d98db1feaeb21863530e6ce.

To render your post, could you please add the missing `index.bib` file?

File: posts/benchmarking_design_decisions/index.qmd [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+to a user that an issue has arisen during code execution [@wickham2019advanced].
+
+Base R ships with functionalities for condition signalling. There is
+`base::stop()` for signalling errors.  Errors refer to cases where something fatal
+happens during code execution so that it halts, for example, when a function is
+run without specifying the required arguments.
+
+```{r}
+try(rnorm())
+```
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:50]:
````suggestion
```{r, error = TRUE}
rnorm()
```
````

File: posts/benchmarking_design_decisions/index.qmd [hunk 20]
```diff
@@ -0,0 +1,266 @@
+---
+title: "Using benchmarking to guide the adoption of dependencies in R packages"
+author:
+  - name: "James Mba Azam"
+    orcid: "https://orcid.org/0000-0001-5782-7330"
+date: last-modified
+categories: [benchmarking, R, cli, R packages, software architecture]
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:7]:
```suggestion
categories: [benchmarking, R, cli, R package, software architecture]
```

File: posts/benchmarking_design_decisions/index.qmd [hunk 21]
```diff
@@ -0,0 +1,266 @@
+---
+title: "Using benchmarking to guide the adoption of dependencies in R packages"
+author:
+  - name: "James Mba Azam"
+    orcid: "https://orcid.org/0000-0001-5782-7330"
+date: last-modified
```
Comment (Bisaloo) [inline posts/benchmarking_design_decisions/index.qmd:6]: The `date` field will need to be updated to a static date before publishing to avoid re-publishing this post if we ever fix a typo or an URL in it.

**Review by Bisaloo — CHANGES_REQUESTED** [review]

File: posts/benchmarking_design_decisions/index.qmd (hunk 21 above)
Comment (jamesmbaazam) [reply posts/benchmarking_design_decisions/index.qmd]: I've fixed that here acf2ede453b5ea607d98db1feaeb21863530e6ce.

**Comment by jamesmbaazam** [comment]
> To render your post, could you please add the missing index.bib file?

I've added it but the rendering CI is still failing because of the `bench` package.

**Review by Bisaloo — APPROVED** [review]

**Comment by Bisaloo** [comment]
I can merge it if you confirm you're ready. General recommendation in the blueprints is maintainer but I agree there has been variability and overcommunication is always a better thing in this situation so thanks for asking :+1:

**Comment by jamesmbaazam** [comment]
I believe it's ready.
